src/control/lib/daos/api/libdaos_stubs.go

```diff
@@ -0,0 +1,116 @@
+//
+// (C) Copyright 2024 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+//go:build test_stubs
+// +build test_stubs
+
+package api
+
+import (
+	"unsafe"
+
+	"github.com/daos-stack/daos/src/control/build"
+	"github.com/daos-stack/daos/src/control/lib/daos"
+)
+
+/*
+#include <daos_errno.h>
+#include <daos_mgmt.h>
+*/
+import "C"
+
+var (
+	daos_init_RC C.int = 0
+)
+
+func daos_init() C.int {
+	return daos_init_RC
+}
+
+func daos_fini() {}
+
+func dc_agent_fini() {}
+
+var (
+	defaultSystemInfo *daos.SystemInfo = &daos.SystemInfo{
+		Name:      build.DefaultSystemName,
+		Provider:  "ofi+tcp",
+		AgentPath: "/does/not/exist",
+		RankURIs: []*daos.RankURI{
+			{Rank: 0, URI: "/does/not/exist"},
+			{Rank: 1, URI: "/does/not/exist"},
+			{Rank: 2, URI: "/does/not/exist"},
+		},
+		AccessPointRankURIs: []*daos.RankURI{
+			{Rank: 0, URI: "/does/not/exist"},
+			{Rank: 1, URI: "/does/not/exist"},
+			{Rank: 2, URI: "/does/not/exist"},
+		},
+	}
+	daos_mgmt_get_sys_info_SystemInfo *daos.SystemInfo = defaultSystemInfo
+	daos_mgmt_get_sys_info_RC         C.int            = 0
+)
+
+func daos_mgmt_get_sys_info(group *C.char, sys_info_out **C.struct_daos_sys_info) C.int {
```

Comment:
I am definitely open to ideas on how to make these better... The main thing that I went around and around on over the past year was how to make most of the logic around the bindings testable without contorting the production code too much in service of the tests.

In another patch that adds a lot more to the bindings, I experimented with an interface-based approach and it wound up feeling pretty gross, with a giant interface definition that could be implemented by either the real thing or a mock. Even if we could somehow break it down into smaller interfaces that were specific to each API call, it seems like we're just shifting the maintenance mess around a bit.

Reply:
I guess the thing that bothers me more than the approach itself is that the layers are a bit jumbled... There's Provider, api, and then a wrapper layer directly above the actual C functions to allow the stubbing. Sometimes Provider calls api, and sometimes it calls the wrapper functions directly. Straightening that relationship out would make me feel a bit better about it. As I was playing around with it, I kept wanting to make api the simple wrapper layer, and do everything else (including the extra init/fini bookkeeping) in the Provider.

Reply:
> There's Provider, api, and then a wrapper layer directly above the actual C functions to allow the stubbing. 

Yeah, those are all artifacts from a couple of different approaches to implementing the bindings. My inclination at this point is to continue down the path that I've started with this PR, i.e. the stubs, and gradually eliminate the other stuff. As I've said, though, I'm open to other ideas about how to do this. I'd like to settle on something we can live with for the long term though, because I don't want to get into mass rewrites later.

Reply:
I'm fine with landing as-is and making incremental improvements to this implementation. As we discussed elsewhere, it would be great to have a more explicit form of dependency injection, but we don't need to have it 100% perfect on the first go-round.